Run serialization tests against upstream types and update v1beta3

kargakis opened the pull request:
Fixes https://github.com/openshift/origin/issues/5483

@deads2k @pmorie @smarterclayton @pweil-

kargakis commented:
@pmorie too

kargakis commented:
[test]

deads2k commented on Godeps/_workspace/src/k8s.io/kubernetes/pkg/api/serialization_test.go:
```diff
@@ -134,17 +136,19 @@ func TestRoundTripTypes(t *testing.T) {
-		// Try a few times, since runTest uses random values.
-		for i := 0; i < *fuzzIters; i++ {
-			doRoundTripTest(kind, t)
+		for version, codec := range map[string]runtime.Codec{"": api.Codec, "v1beta3": v1beta3.Codec, "v1": v1.Codec} {
```
Given the way that UPSTREAM likes to test by re-running the whole bucket with a different version, I don't think we'll be able to get the structure of this changed moved up.

What about updating the `serialization_test.go` in our `pkg/api` to not exclude kube types?

kargakis replied:
> What about updating the serialization_test.go in our pkg/api to not exclude kube types?

I think this is plausible. Will try it.

kargakis replied:
@deads2k how about now?

kargakis commented:
Btw, I see that all TypeMeta in v1beta3 need to be updated to unversioned.TypeMeta:)

deads2k commented on pkg/api/serialization_test.go:
```diff
@@ -405,9 +406,15 @@ func TestSpecificKind(t *testing.T) {
 }
 
+// Keep this in sync with the respective upstream set
+var nonInternalRoundTrippableTypes = sets.NewString("List", "ListOptions", "PodExecOptions", "PodAttachOptions")
```
Sorry, this isn't obvious to me.  These are somehow broken?  Also, anything that needs to stay in sync should have a unit test to make sure it stays in sync.

kargakis replied:
It isn't obvious to me either, so I located the PRs which added those upstream:
PodAttachOptions: https://github.com/kubernetes/kubernetes/pull/12119
PodExecOptions: https://github.com/kubernetes/kubernetes/pull/6869
ListOptions: https://github.com/kubernetes/kubernetes/pull/5763
List: https://github.com/kubernetes/kubernetes/pull/2786

@smarterclayton may have a word here

pmorie replied:
Should we have test for each of these in `compatibility_test.go`?

kargakis commented:
I have removed status which now is an unversioned object and added PodAttachOptions to v1beta3.

Object that still need to be updated in v1beta3:
* PodLogOptions
* Nodes
* PV
* Endpoints
* Services
* PodStatusResult

Based on the massive error returned from the serialization test. We need to make it more readable.

kargakis commented:
Have boiled down serialization failures to Pods

kargakis commented:
Pod serialization fixed in https://github.com/kargakis/origin/commit/5e64f74aab47234b4967379621cdb093208caf01

this needs review from @pmorie and @pweil-

liggitt commented on pkg/api/serialization_test.go:
```diff
@@ -415,9 +416,15 @@ func TestSpecificKind(t *testing.T) {
 func TestTypes(t *testing.T) {
 	for kind, reflectType := range api.Scheme.KnownTypes("") {
-		if !strings.Contains(reflectType.PkgPath(), "/origin/") {
+		if !strings.Contains(reflectType.PkgPath(), "/origin/") && reflectType.PkgPath() != "k8s.io/kubernetes/pkg/api" {
```
why is this needed? if the pkg path does equal `"k8s.io/kubernetes/pkg/api"`, the first condition would already skip it...

kargakis replied:
This is AND not OR...

liggitt replied:
ah. Can we update the name of this test if the intent is to test serialized k8s types?

kargakis replied:
I like the present name. How about adding godoc?

On Fri, Oct 30, 2015 at 3:10 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/api/serialization_test.go
> <https://github.com/openshift/origin/pull/5485#discussion_r43505303>:
>
> >  func TestTypes(t *testing.T) {
> >  	for kind, reflectType := range api.Scheme.KnownTypes("") {
> > -		if !strings.Contains(reflectType.PkgPath(), "/origin/") {
> > +		if !strings.Contains(reflectType.PkgPath(), "/origin/") && reflectType.PkgPath() != "k8s.io/kubernetes/pkg/api" {
>
> ah. Can we update the name of this test if the intent is to test
> serialized k8s types?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5485/files#r43505303>.
>

liggitt replied:
nevermind, I apparently can't interpret boolean statements today

liggitt commented:
can we get a hand crafted test specifically around pod spec in `pkg/api/compatibility_test.go` to make sure an old v1beta3 serialization decodes to what we expect? @pmorie might have a good one

kargakis commented:
I finally got a clean run locally

deads2k commented:
I think I'm willing to take an issue to figure out this: https://github.com/openshift/origin/pull/5485/files#diff-394377845c8dfd9eec363c059cdde354R420, squash the `UPSTREAM: <carry>'s` down to one and merge to unstick @pmorie's pull.

kargakis commented:
We still need @pmorie to review https://github.com/kargakis/origin/commit/9a3bcf6c2c96e600d1ad387061cd062000a6f14b

kargakis commented:
> I think I'm willing to take an issue to figure out this: https://github.com/openshift/origin/pull/5485/files#diff-394377845c8dfd9eec363c059cdde354R420, squash the UPSTREAM: <carry>'s down to one and merge to unstick @pmorie's pull.

Opened https://github.com/openshift/origin/issues/5529

deads2k commented:
@pmorie I think the last sign off is yours: 

> We still need @pmorie to review kargakis@9a3bcf6

pmorie commented:
@deads2k I have one comment on the conversion, LGTM otherwise.

kargakis commented:
re[test]

kargakis commented:
new-app flake

[test] again

deads2k commented:
lgtm [merge]

pmorie commented:
Also LGTM, thanks for the clarification on the conversion.
